### DISCUSSION TITLE
ESP32 with CAN

### DISCUSSION THREAD

**Discussion by stefanheinrichsen** [open3e/open3e#29] [Ideas]
Did anyone here tried to use a ESP32 with CAN transceiver already? Maybe even with esphome firmware (or mikropyhton to at least stay a bit in the framework used here)?

While an pi is nice from the setup perspective it is a bit greedy when it comes to power consumption...

(now I've written everything in english but noticed the other discussions are in german.... Dabei wäre mir deutsch hier auch lieber :-D)

**Comment by jwende** [comment]
Not yet … but interesting idea. I didn‘t found any uds or isotp modules for ESP so far. 
If you use basic modules and vcan tunnels (#cannelloni) you delegate most of the work to the server side.
Cheers.
J.

**Reply by stefanheinrichsen** [reply]
Last time I really worked with CAN was ~25 years ago... so excuse my ignorance: What is "uds" and "isotp"? My first idea was simply use [this](https://esphome.io/components/canbus.html) and then create a esphome yaml file. On the linked pages they refere to a board based on a mcp2515. Would this work or are the viessmann devices using some special CAN setup which is not suppored by this mcp2515?

**Reply by jwende** [reply]
Someone at Viessmann decided for a high level CAN protocol that is widely used in automotive …
Please see here for an overview of the stack: https://de.mathworks.com/products/connections/product_detail/uds-blockset.html.
So it would be an interesting task to use the protocol using plain base CAN Bus messages.

**Reply by jwende** [reply]
So the physical layer could be built easily using ESP and the CAN wire modules …

**Reply by MyHomeMyData** [reply]
As Jörg already mentioned, [cannelloni](https://github.com/mguentner/cannelloni) could be the best solution - a [ESP32 implementation](https://github.com/PhilippFux/ESP32_CAN_Interface) seems available.

If you get this running on the ESP32 you could use cannelloni on your (Linux based) home automation system to receive the CAN data on virtual CAN vcan0.

Then use open3e with vcan0 as you would do with a physical CAN interface.

**Reply by philippoo66** [reply]
> it would be an interesting task to use the protocol using plain base CAN Bus messages

more time-consuming than interesting I think. it would mean to re-invent the wheel. with isotp and udsoncan exactly this work was already done, but utilizing the 'can kernel' which seems not to be available on ESP...

open3e uses only a very small subset of UDS - read and write data by identifier. for this isotp is almost enough. Perhaps it is possible to copy the according code of isotp and get it working with MicroPython on ESP. But I think the canneloni way is the far better one - no stupid work, just fun ;-)

**Reply by stefanheinrichsen** [reply]
Most probably I have a few days spare time before and even more in the beginning of 2024. Will try to build something and come back here with learnings.

**Reply by M4n197** [reply]
@stefanheinrichsen Ever since open3e and @philippoo66 revealed the secrets of UDS to me, I've been reading/writing raw CAN frames on an ESP32 with ESPHome.  I'm using one of the popular MCP2515 boards, but that may not be necessary. 
It gets the job done and works well so far. Of course, this isn't nearly as polished or compelling as open3e, but it satisfies my needs.

**Reply by Styx85** [reply]
@M4n197 Can you post your yaml how you did it in esphome?

**Reply by M4n197** [reply]
> @M4n197 Can you post your yaml how you did it in esphome?

Sure, I hope it helps, even if it got pretty messy over time:

[esp32-e380ca-canlistener_26112023.yaml.txt](https://github.com/abnoname/open3e/files/13467390/esp32-e380ca-canlistener_26112023.yaml.txt)

Requests are initiated using the time-component:
```
time:
- platform: homeassistant
  id: homeassistant_time
  on_time:
    
    #1664: O3EInt8(1, "ElectricalEnergyStorageStateOfCharge"),                          -> DID: 1664 = 0x0680
    - seconds: /30
      then: 
        lambda: |-
          std::vector< uint8_t > data0{ 0x03, 0x22, 0x06, 0x80, 0x00, 0x00, 0x00, 0x00 };
          id(cb)->send_data(0x680, 0, data0);
```
Correctly reading multi-frame messages took some tinkering, because you can't tell to which DID the received Consecutive Frames (CF) belong. (But you know of course what you've requested.) So I tried to interlock it when detecting the First Frame (FF). This works so far and I did not have any errors interpreting the data anymore. 

When Flow-Control is necessary, I inserted a delay, because some CF's were missing. Also CF's sometimes seem to come out of order. 

Correctly reading 32bit-words splitted through CF's isn't done right here, because most of the time I'm lucky and don't need the full word to correctly interpret the data. However, this could be added.

**Reply by philippoo66** [reply]
interesting! I understand nearly nothing :-))

but EVIL - you posted your wifi_password in the ww internet ;-)

**Reply by M4n197** [reply]
> interesting! I understand nearly nothing :-))

Me too! ^^
The yaml way of configuring this stuff is quite fast and powerful. One can integrate custom c++ code using the "lambdas". However doing all this in a single file and having effectively two syntaxes mixed isn't so nice. And doing it the way I've tried is not professional by any means. I hope this might help @Styx85 or @stefanheinrichsen a little bit. Writing a custom component would be the proper way.

> but EVIL - you posted your wifi_password in the ww internet ;-)

The AP-password? Yep, I missed it - big Oopsi! 🤦‍♂️  But luckily this is nothing critical and already fixed. Anyway thanks for the hint.

**Reply by redarflap** [reply]
Does by any chance somebody know if there are any parallels to the older ex-Köb Pyromat ECO models (Viessmann labeled model from 2012)

I am listening to the CAN bus with an ESP32, but I'm only getting spammed with empty 0x701 messages at 125kBit/s.
Do those systems generally only answer to requests or should they broadcast something?

**Reply by philippoo66** [reply]
are you sure about the baudrate? The current Viessmanns do 250kBd

**Reply by philippoo66** [reply]
701 is Node/Life Guarding / Heartbeat of node 01 but should contain at least 1 byte data. Perhaps you need to request something to get something?! or set operational?! How much do you know about CAN and higher layer protocols?

**Reply by redarflap** [reply]
I'm pretty sure it's 125kBd because i wrote a test programm that tried all the settings and 125kBd was the only one where it picked up something.
I really didn't know about the existence of higher layer protocols on top of CAN before reading this discussion, but now that you mention 0x701 being a heartbeat packet that would totally make sense.

However, in some places, there is talk about a data/visualisation feature described as a paid unlockable feature, but i'm not sure if that's related to the CAN bus directly on the Pyromat ECO or something external like that ECOTRONIC controller that comes with an RS232 port.

Thinking further about it, this little idea becomes more and more complex. The heater is a blackbox to me, I have no reference to any communication over this interface and therefore not even the slightest idea how to start. I guess even if i created a working ISO-TP solution, that wouldn't yield anything since i still wouldn't know how to talk to the damn thing.

I guess I'll contact Viessmann support, maybe they can give me some hints to clear my mind about what's even possible.


However, doing further research, i found this little project for esp32, which may be interesting:
https://github.com/bri3d/esp32-isotp-ble-bridge

**Reply by philippoo66** [reply]
I think the ISO-TP thing came with OneBase/E3. Have you tried 'simple' CANopen SDO service already? Or just set that thing in CANopen operational state - perhaps it starts sending PDOs then...

**Reply by redarflap** [reply]
I've tried no such thing, but after you mentioning CANopen, i've found the following:
https://www.embedded-communication.com/en/canopen/english-canopen-bootup-message/

The description of the bootup message with ID 0x701 matches exactly my situation. Apparently the 1 byte of data is 0x00, so that might explain why i didn't see any data. It probably has been printed to the console as a white space. Actually i never checked the actual length, only the content.

 I will try to set something up in the next days. Thanks a lot!

**Reply by philippoo66** [reply]
to my experience (and to CANopen rules) the boot-up message gets sent only once after bootup. Afterwards (after initialisation) the node should automatically enter the pre-operational state

![grafik](https://github.com/open3e/open3e/assets/122479122/8ccb4266-6468-4c33-96c3-08b7015ca677)

But since nobody knows if the device follows CANopen rules, noboy knows...

To get a little bit further you might try to read out certain basic objects to see _if_ the device _is_ CANopen. I'm attaching a 'poster' showing the CANopen objects / object ranges. 

[CANopen_poster.pdf](https://github.com/open3e/open3e/files/13692075/CANopen_poster.pdf)

to initiate a SDO upload request, a frame with COB-ID 600h+node-id 
byte 0 : 40h
byte 1,2 : object index lsb..msb
byte 3 : sub-index 
byte 4..7 : reseverd
needs to be sent.

The COB-ID of the response frame usually would be 580h+node-id
The service will work only at least at pre-operational state 

e.g. the Identity Objekt 1018h is mandatory with CANopen devices. At sub-index 1 the Vedor-ID is noted.
`cansend can0 601#4018100100000000`
would be the command with can-utils, I don't know how to do with 'ESP can' ;-)

if you get a value (bytes 4..7, lsb..msb!), you can look it up here https://www.can-cia.org/services/canopen-vendor-id/ Then it makes sense to care for further objects.

**Reply by philippoo66** [reply]
p.s. to start remote node id 01 (set it operational - make it transferring it's PDOs) the command
`cansend can0 000#0101`
rsp. `cansend can0 000#0101000000000000`
needs to get sent.
To set it pre-operstional (probably necessary) send `cansend can0 000#8001`

![grafik](https://github.com/open3e/open3e/assets/122479122/13aab46e-4445-4463-8f61-085be5a4c990)

**Reply by redarflap** [reply]
Thanks a lot for all those inputs, they will definitely be handy when I get to trying it. In the meantime I also read alot about it myself.

I think you may be wrong about the bootup-message only being sent once. That may be true if there are other devices on the bus. However according to this website (https://www.embedded-communication.com/en/canopen/english-canopen-bootup-message/), if the bootup-message never gets acknowledged, the device just starts spamming it indefinitely.
See section "Bootup of isolated CANopen device".

Wen I find some time, I will start digging into https://github.com/sicrisembay/CANopenNode_ESP32

**Reply by redarflap** [reply]
So, i made some progress :)
With the help of [CANopenNode_ESP32](https://github.com/sicrisembay/CANopenNode_ESP32), i managed to set up the CANopen communication. The ESP32 is operational and is getting flooded with RPDO messages:

```
I (9498) CO_driver: CANRX id: 0x23b, dlc: 8, data: [0 0 0 0 0 0 0 0]
I (9499) CO_driver: CANRX id: 0x33b, dlc: 8, data: [3 0 0 0 0 0 1 0]
I (9501) CO_driver: CANRX id: 0x43b, dlc: 8, data: [6 0 0 0 ae 1 0 0]
I (9508) CO_driver: CANRX id: 0x53b, dlc: 8, data: [9 0 0 0 0 0 0 0]
I (9598) CO_driver: CANRX id: 0x23b, dlc: 8, data: [c 0 0 0 79 0 67 0]
I (9599) CO_driver: CANRX id: 0x33b, dlc: 8, data: [f 0 1 0 80 0 1 0]
I (9601) CO_driver: CANRX id: 0x43b, dlc: 8, data: [12 0 6 0 0 0 0 0]
I (9608) CO_driver: CANRX id: 0x53b, dlc: 8, data: [15 0 0 0 0 0 0 0]
I (9698) CO_driver: CANRX id: 0x23b, dlc: 8, data: [18 0 0 0 0 0 0 0]
I (9699) CO_driver: CANRX id: 0x33b, dlc: 8, data: [1b 0 0 0 0 0 0 0]
I (9701) CO_driver: CANRX id: 0x43b, dlc: 8, data: [1e 0 1 0 95 1 a1 0]
I (9708) CO_driver: CANRX id: 0x53b, dlc: 8, data: [21 0 e8 3 de 1 f9 1]
I (9798) CO_driver: CANRX id: 0x23b, dlc: 8, data: [24 0 7b 0 dc 2 c2 2]
I (9798) CO_driver: CANRX id: 0x33b, dlc: 8, data: [27 0 b9 2 e8 3 e8 3]
...
```

Just for clarity, the ESP32 receiving RPDO messages, means the other side is requesting data (TPDO) right? Somehow I didn't find a clear explanation of this.

Tomorrow I will try to fetch the vendor-id using SDO.

**Reply by philippoo66** [reply]
the 'RX' indicates that a message is received (by the ESP CAN). A RPDO is never transmitted. And since you are not the consumer of the RPDOs these messages are TPDOs (if they are PDOs). But that doesn't really matter, it's CAN data frames - means that data get transmitted. That's fine because now we see that the system 'is alive'.

As soon as you are ready to try to do a SDO scan or are willing to find out if it's possible to read out the PDO mapping let me know and I will give support. 

ps. if some node transmits PDOs it means that _this_ device is operational. The ESP CAN interface itselfe is no device which will assume such kind of NMT state. It only receives frames and it may transmit frames if you or a program you wrote make it transmitting. 'operational' describes a status where a device is acting according to the rules described by its PDO mapping and PDO communication parameters. This is much to read to understand (but perhaps not completely necessary). The CANopen poster I posted might give you a little bit of an idea...

but seeing CAN frames being transmitted it does not mean that this must be CANopen. CANopen consists of multiple of the higher layers and perhaps what we see is simply CAN ;-)

**Reply by redarflap** [reply]
Okay, got it :)

After a lot of trial and error with the above mentioned CANOpenNode ESP32 component, I managed to set up two ESP32 which were able to read each others device ids (which i set in their object dictionaries). 

However when trying the same on the oven, alltough I do get answers, the vendor ID seems to be set to 0x1.

```
I (91497) CO_driver: CANTX id: 0x601, dlc: 8, data: [40 18 10 1 0 0 0 0]
I (91504) CO_driver: CANRX id: 0x581, dlc: 8, data: [43 18 10 1 1 0 0 0]    Sub-Idx1: VendorId = 1
I (91518) CO_driver: CANTX id: 0x601, dlc: 8, data: [40 18 10 2 0 0 0 0]     
I (91524) CO_driver: CANRX id: 0x581, dlc: 8, data: [43 18 10 2 0 0 0 0]    Sub-Idx2: ProductCode = 0
I (91538) CO_driver: CANTX id: 0x601, dlc: 8, data: [40 18 10 3 0 0 0 0]     
I (91544) CO_driver: CANRX id: 0x581, dlc: 8, data: [43 18 10 3 0 0 0 0]    Sub-Idx3: RevNr = 0
I (91559) CO_driver: CANTX id: 0x601, dlc: 8, data: [40 18 10 4 0 0 0 0]      
I (91564) CO_driver: CANRX id: 0x581, dlc: 8, data: [43 18 10 4 0 0 0 0]    Sub-Idx4: SerialNr = 0
```


However, i am currently looking at that PDO? data that is being broadcasted, and i think i have located some interesting adresses. For example I think I have located the RPM of the ovens ventilation system. I still need to come up with a better way of capturing and visualising the data, as the manual capturing and comparison of the monitoring terminal is tedious.

**Reply by philippoo66** [reply]
hey! at least the oven serves SDOs. And you manged to read them out - great! That's nice to know, so it looks a little bit like CANopen. Even the implementation seems not to be too 'straight' ;-)

yap, finding out now what values get transferred is poking around in the fog without any information... With open3e we had ViGuide at least...

you might try to read out the PDO mapping parameters, but if they are there and filled, they would only tell you which SDOs are mapped into the PDOs, but as long as you have no SDO dictionary, that wouldn't help further anyway....

Do you know for what purpose this CAN interface was meant? The WAGO gateway did not exist that time, did it? What else might have been meant to get connected?

**Reply by redarflap** [reply]
The only information that i found about the interface is, that it allows to connect an external controller (Ecotronic) which supposedly allows to control many of the ovens parameters and the controller allows to forward operational variables to other systems via RS232.

Would reading the PDO mapping parameters get me information about data types or size? That would already be a big help.

**Reply by philippoo66** [reply]
_if_ the mapping is 'noted' according to CANopen, it would give you info regarding the data size (byte (bit) len), not the type/interpretation. _if_ there would be any data mapped which are not manufacturer specific but according to some device profile standard, you would know what the data are. but I'm not aware of any CANopen 'oven device profile' standard.

Do you know the PDO mapping mechanism? The CANopen poster gives a brief impression. The object mapping of TPDOs itself would be found at 1A00+ Perhaps it would make sense to give that a try.

Is it possible to get any (technical) information regarding that Ecotronic?

**Reply by redarflap** [reply]
I couldn't find much information about the Ecotronic. About reading out the data sizes, that could be handy.
However maybe that's not even necessary, because I'm already making good progress.

I wrote a small algorithm that keeps track of the previous broadcasts and only prints changes between them.
Then i recorded several sessions while modifying different parameters/states of the oven.
Doing  so I managed to identify some values in the data stream.

- Ventilation RPM
- Clock (hh:mm:ss)
- Rest-O2 setting (%) 
- Oven target temperature (°)

All the values so far are broadcast as unsigned 16bit integer format.
Temperatures are sent in 10th degrees, percentages in promille.

When I re-format the data in my program to show decimals and record a full heating session, I think I should be able to find all the relevant temperatures as well. :-)

Thanks so much for your support.

**Reply by philippoo66** [reply]
Nice to hear that you are doing progress! And nice to hear that I could help a little bit!

Do I understand right that the infos you already identified are all coming with these 4 COB-IDs 23b, 33b, 43b, 53b? And that there are more data in there which you intend to identify either?

These COB-IDs are not standard (pre-defined) for CANopen TPDOs, so again we are not sure what is deal with. If you 'feel boared' it might be interesing to do a SDO scan. Especially ranges 1xyzh, 2xyzh and 6xyzh. 1... is communication profile area in CANopen. At 2... and 6... you might find 'datapoints'/parameters which are not cyclicly posted but might be interesting, too. Probabliy you might find some settings there what could also be interesting...

**Reply by redarflap** [reply]
Yes, the infos all come with those 4 COB-IDs. It sends data from a bunch of offsets (the first two bytes of data representing the offset). And once it's through the list, it starts with the next cycle right away.
I then noticed that the remaining 6 bytes apparently always resemble 3 values of 16-bit integers. After sorting by offset and printing only values that changed, it was straightforward to find pretty much all the values i'm interested in, by comparing with the values shown on the heaters user interface.

If you're interested i appended the raw output of one full cycle.

Maybe i will still do a full SDO scan, but first I need to clean up the code, since i hacked it all together pretty quickly. I'll try to do the initial "handshake" myself so the oven starts sending data, since that whole CANopenNode stack i am using now is pretty convoluted and unnecessary for my case, since i really only needed that initial handshake.

The next step is to send the data to my MQTT gateway via LoRa.
I was thinking about powering the ESP32 through the 24V that are provided by the CAN bus. Do you think that's a bad idea? I know it's supposed to power only galvanically isolated transceivers. But the whole ESP32 doesn't really consume more power than 2-3 transceivers would (still need to measure the peak current).

[pyromat_raw.txt](https://github.com/open3e/open3e/files/13759002/pyromat_raw.txt)

**Comment by philippoo66** [comment]
@maromme continuing here to move the discussion off issues... ([started here](https://github.com/open3e/open3e/issues/3#issuecomment-1888934709))
```
   0  1  2  3  4  5  6  7
0x 10 0C 62 01 47 ED FF CB
0x 21 FF 8F 00 EC FF 00 55
```
> the payload itselfe should be then 62 01 47 ED FF CB FF 8F 00 EC FF 00 55 where the last 55 is padding for to get 12 byte

minimally different: the data length 0C of the payload includes SID (0x62) which always (in case of success) is SID of the request +40h, and after this the DID repeated: 0147h = 327. So the 'pure data' is `ED FF CB FF 8F 00 EC FF 00` (always len-3). These are 4x 2 bytes values (-1.9°C, -5.3°C,  14.3°C, -2.0°C) and one sensor status byte (00 = no error). 

The padding bytes are never included in UDS data length, only in the data length code (DLC) of the CAN frame itself.

**Reply by maromme** [reply]
@philippoo66: wow thanks for your replay. sorry for beeing late with an answer, but well time...
I made a running setup. going to document it and then poste it. but that might take some time.

**Reply by maromme** [reply]
here you can find my setup: https://github.com/maromme/esphome-vitoair/
